source_modelling/sources.py

```diff
@@ -0,0 +1,714 @@
+    ) -> np.ndarray:
+        """Convert fault coordinates to global coordinates.
+
+        See global_coordinates_to_fault_coordinates for a description of fault
+        coordinates.
+
+        Parameters
+        ----------
+        fault_coordinates : np.ndarray
+            The fault coordinates of the point.
+
+        Returns
+        -------
+        np.ndarray
+            The global coordinates (lat, lon, depth) for this point.
+        """
+
+        # the right edges as a cumulative proportion of the fault length (e.g. [0.1, ..., 0.8])
+        right_edges = self.lengths.cumsum() / self.length
+        fault_segment_index = np.searchsorted(right_edges, fault_coordinates[0])
+        left_proportion = (
+            right_edges[fault_segment_index - 1] if fault_segment_index > 0 else 0
+        )
+        right_proportion = (
+            right_edges[fault_segment_index + 1]
+            if fault_segment_index < len(right_edges) - 1
+            else 1
+        )
+        segment_proportion = (fault_coordinates[0] - left_proportion) / (
+            right_proportion - left_proportion
+        )
+        return self.planes[
+            fault_segment_index
+        ].fault_coordinates_to_wgs_depth_coordinates(
+            np.array([segment_proportion, fault_coordinates[1]])
+        )
+
+
+class HasCoordinates(Protocol):
+    """Type definition for a source with local coordinates."""
+
+    def fault_coordinates_to_wgs_depth_coordinates(
+        self,
+        fault_coordinates: np.ndarray,
+    ) -> np.ndarray: ...
+
+    def wgs_depth_coordinates_to_fault_coordinates(
+        self,
+        fault_coordinates: np.ndarray,
+    ) -> np.ndarray: ...
+
+
+def closest_point_between_sources(
+    source_a: HasCoordinates, source_b: HasCoordinates
+) -> tuple[np.ndarray, np.ndarray]:
+    """Find the closest point between two sources that have local coordinates.
+
+    Parameters
+    ----------
+    source_a : HasCoordinates
+        The first source. Must have a two-dimensional fault coordinate system.
+    source_b : HasCoordinates
+        The first source. Must have a two-dimensional fault coordinate system.
+
+    Raises
+    ------
+    ValueError
+        Raised when we are unable to converge on the closest points between sources.
+
+    Returns
+    -------
+    source_a_coordinates : np.ndarray
+        The source-local coordinates of the closest point on source a.
+    source_b_coordinates : np.ndarray
+        The source-local coordinates of the closest point on source b.
+    """
+
+    def fault_coordinate_distance(fault_coordinates: np.ndarray) -> float:
+        source_a_global_coordinates = (
+            source_a.fault_coordinates_to_wgs_depth_coordinates(fault_coordinates[:2])
+        )
+        source_b_global_coordinates = (
+            source_b.fault_coordinates_to_wgs_depth_coordinates(fault_coordinates[2:])
+        )
+        return coordinates.distance_between_wgs_depth_coordinates(
```

Comment:
This is a function that either needs to be in qcore or I can copy it into this repo. I think a distance function between coordinates (like `ll_dist` but vectorised, more accurate, etc) is probably welcome in qcore?

Reply:
I agree this could be welcomed into qcore if it's going to be used in multiple repos